keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+### Out-of-tree Implementation
+
+We could implement the NodeMaintenance or Evacuation API out-of-tree first as a CRD.
+
+The KEP aims to solve graceful termination of any pod in the cluster. This is not possible with a
+3rd party CRD as we need an integration with core components.
+
+- We would like to solve the lifecycle of static pods during a node maintenance. This means that
+  static pods should be terminated during the drain according to `drainPlan`, and they should stay
+  terminated after the kubelet restart if the node is still under maintenance. This requires
+  integration with kubelet. See [kubelet: Static Pods](#kubelet-static-pods) for more details.
+- We would like to improve the Graceful Node Shutdown feature. Terminating pods via NodeMaintenance
+  will improve application safety and availability. It will also improve the reliability of the
+  Graceful Node Shutdown feature. However, this also requires the kubelet to interact with a
+  NodeMaintenance. See [kubelet](#kubelet) and
+  [kubelet: Graceful Node Shutdown](#kubelet-graceful-node-shutdown) for more details.
+- We would like to also solve the lifecycle of DaemonSet pods during the NodeMaintenance. Usually
+  these pods run important or critical services. These should be terminated at the right time
+  during the node drain. To solve this, integration with NodeMaintenance is required. See
+  [DaemonSet Controller](#daemonset-controller) for more details.
+
+Also, one of the disadvantages of using a CRD is that it would be more difficult to get real-word
```

Comment:
One of the reasons for introducing CRDs was to allow authors to iterate faster and provide feedback which can:
1. improve the core k8s;
2. allow to make an informed decision whether a particular problems is worth integrating in core or not. 

So far, the several different projects you've mentioned in this document were able to achieve the functionality you're proposing here outside of core.

Reply:
I see, I need to emphasize this better in the goals. The kubelet integration is an important part of this KEP. 

The other projects are mostly replicating the kubectl drain logic in a controller with some extra. If this were the only part of the KEP, then I would agree that a CRD would be preferable.

Reply:
(I think this strengthens the case for prototyping with a CRD and then switching in-tree when we add the kubelet integration).

Reply:
Correct, the whole point of out-of-tree experimentation is to shape up the API and identify what has to be done in the core, that is not possible today, when using an external solution.